pkg/oracle/response.go

```diff
@@ -0,0 +1,155 @@
+package oracle
+
+import (
+	"encoding/hex"
+	"errors"
+	gio "io"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/fee"
+	"github.com/nspcc-dev/neo-go/pkg/core/native"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/manifest"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/emit"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+	"go.uber.org/zap"
+)
+
+var oracleInvoc []byte
+
+func init() {
+	w := io.NewBufBinWriter()
+	emit.Int(w.BinWriter, 0)
+	emit.Opcode(w.BinWriter, opcode.PACK)
+	emit.String(w.BinWriter, manifest.MethodVerify)
+	if w.Err != nil {
+		panic(w.Err)
+	}
+	oracleInvoc = w.Bytes()
+}
+
+func (o *Oracle) getResponse(reqID uint64) *incompleteTx {
+	o.respMtx.Lock()
+	defer o.respMtx.Unlock()
+	incTx, ok := o.responses[reqID]
+	if !ok {
+		incTx = newIncompleteTx()
+		o.responses[reqID] = incTx
+	}
+	return incTx
+}
+
+// AddResponse processes oracle response from node pub.
+// sig is response transaction signature.
+func (o *Oracle) AddResponse(pub *keys.PublicKey, reqID uint64, txSig []byte) {
+	incTx := o.getResponse(reqID)
+
+	incTx.Lock()
+	tx := incTx.tx
+	if tx != nil {
+		ok := pub.Verify(txSig, tx.GetSignedHash().BytesBE())
+		if !ok {
+			o.Log.Debug("invalid response signature",
+				zap.String("pub", hex.EncodeToString(pub.Bytes())))
+			incTx.Unlock()
+			return
+		}
+	}
+	incTx.addResponse(pub, txSig)
+	ready := incTx.finalize(o.getOracleNodes())
+	incTx.Unlock()
+
+	if ready {
+		o.OnTransaction(tx)
+	}
+}
+
+func readResponse(rc gio.ReadCloser, limit int) ([]byte, error) {
+	defer rc.Close()
+
+	buf := make([]byte, limit+1)
+	n, err := gio.ReadFull(rc, buf)
+	if err == gio.ErrUnexpectedEOF && n <= limit {
+		return buf[:n], nil
+	}
+	return nil, errors.New("too big response")
+}
+
+// CreateResponseTx creates unsigned oracle response transaction.
+func (o *Oracle) CreateResponseTx(gasForResponse int64, resp *transaction.OracleResponse) (*transaction.Transaction, error) {
+	tx := transaction.New(o.Network, native.GetOracleResponseScript(), 0)
+	tx.Nonce = 0
+	tx.ValidUntilBlock = o.Chain.BlockHeight() + transaction.MaxValidUntilBlockIncrement
+	tx.Attributes = []transaction.Attribute{{
+		Type:  transaction.OracleResponseT,
+		Value: resp,
+	}}
+
+	oracleSignContract := o.getOracleSignContract()
+	tx.Signers = []transaction.Signer{
+		{
+			Account: o.oracleHash,
+			Scopes:  transaction.None,
+		},
+		{
+			Account:          hash.Hash160(oracleSignContract),
+			Scopes:           transaction.None,
+			AllowedContracts: []util.Uint160{o.oracleHash},
+		},
+	}
+	tx.Scripts = []transaction.Witness{
+		{
+			InvocationScript: oracleInvoc,
```

Comment:
It won't work as provided in `neo-modules` PR. There is an issue with calling `verify` on native contracts (because it is done by offset). Isn't a problem, because this doesn't influence transaction hash.